src/model/ve_bus_state.rs

```
@@ -0,0 +1,105 @@
+use crate::bit_reader::BitReader;
+use crate::err::*;
+use num_enum::TryFromPrimitive;
+
+use super::mode::Mode;
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
+pub struct VeBusState {
+    pub mode: Mode,
+    pub error: u8,
+    pub battery_voltage_v: f32,
+    pub battery_current_a: f32,
+    pub ac_in_state: AcInState,
+    pub ac_in_power_w: f32,
+    pub ac_out_power_w: f32,
+    pub alarm: AlarmNotification,
+    pub battery_temperature_c: f32,
+    pub soc_percent: f32,
+}
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd, TryFromPrimitive)]
+#[repr(u8)]
+pub enum AcInState {
+    AcIn1 = 0,
+    AcIn2 = 1,
+    NotConnected = 2,
+    Unknown = 3,
+}
+
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd, TryFromPrimitive)]
+#[repr(u8)]
+pub enum AlarmNotification {
+    NoAlarm = 0,
+    Warning = 1,
+    Alarm = 2,
+}
+
+impl VeBusState {
+    pub(crate) fn parse(payload: &[u8]) -> Result<Self> {
+        let mut reader = BitReader::new(payload);
+
+        let mode = Mode::try_from(reader.read_unsigned_int(8)?)?;
+        let error = reader.read_unsigned_int(8)? as u8;
+        let battery_current_a = (reader.read_signed_int(16)? as f32) / 10.0;
+        let battery_voltage_v = (reader.read_unsigned_int(14)? as f32) / 100.0;
+        let ac_in_state =
```

Please make the parse fail on invalid input data, something like:

```rust
.ok_or(Error::InvalidAcInState)?
```